vimode/src/cmds/edit.c

```diff
@@ -162,14 +163,14 @@ void cmd_del_word_left(CmdContext *c, CmdParams *p)
 void cmd_undo(CmdContext *c, CmdParams *p)
 {
 	gint i;
-	gint pos = SSM(p->sci, SCI_GETCURRENTPOS, 0, 0);
+	start_undo(c);
 	for (i = 0; i < p->num; i++)
 	{
 		if (!SSM(p->sci, SCI_CANUNDO, 0, 0))
 			break;
 		SSM(p->sci, SCI_UNDO, 0, 0);
 	}
```

Comment:
Shouldn't `goto_nonempty()` be used here to go to the first non-white character on a line? vim seems to behave this way, even though in a bit inconsistent manner - when you perform `dd`, undo goes to the first non-white character while when you use `2dd`, it goes where the cursor was before.

If `goto_nonempty()` is used, it should only be run when some actual undo happened and also when the cursor is at the beginning of a line so it doesn't jump somewhere else when you e.g. delete a word in the middle of a sentence.

Reply:
OK, I made these changes and I am now using `goto_nonempty()`.
This is probably the simplest way to approach VIM's behavior.

Reply:
Well, what I had in mind was something like (beware, totally untested!!!):
```C
void cmd_undo(CmdContext *c, CmdParams *p)
{
	gboolean undo_performed = FALSE;
	gint pos;
 	gint i;

 	undo_start(c);

 	for (i = 0; i < p->num; i++)
 	{
 		if (!SSM(p->sci, SCI_CANUNDO, 0, 0))
 			break;
 		SSM(p->sci, SCI_UNDO, 0, 0);
		undo_performed = TRUE;  // <--- this is added
 	}

	undo_end(c);

	// the code below is added
	pos = SSM(ctx->sci, SCI_GETCURRENTPOS, 0, 0);

	if (undo_performed && is_start_of_line(p->sci, pos))
		goto_nonempty(p->sci, pos, FALSE);
}
```

without the need to pass anything extra in the context. Am I missing something? Since I didn't run or even compile it, it's quite possible.

`excmd_undo()` could just call this function then.

Reply:
OK, I tried to refactor the code [63b7e25](https://github.com/geany/geany-plugins/pull/1328/commits/63b7e25242e485448c81185eadabab38d7a65b52).
But `excmd_undo` cannot simply call `cmd_undo` because of the use of `CmdParams` parameter.
So I moved this code into a new `undo_apply` function which is called by `cmd_undo` and `excmd_undo`.